viewing_party/party.py

```diff
@@ -1,23 +1,276 @@
 
+# calculates the average rating of every movie the user has watched
+def get_watched_avg_rating(user_data):
+    rating = 0
+    for movie in user_data["watched"]:
+        rating += movie["rating"]
+    if rating == 0:
+        return 0
+    else:
+        avg_rating = rating / len(user_data["watched"])
+        return avg_rating
+
+# finds the user's most frequently watched genre
+def get_most_watched_genre(user_data):
+    if len(user_data["watched"]) == 0:
+        return None
+    horror_counter = 0
+    fantasy_counter = 0
+    action_counter = 0
+    intrigue_counter = 0
+    for movie in user_data["watched"]:
+        if movie["genre"] == "Horror":
+            horror_counter += 1
+        elif movie["genre"] == "Fantasy":
+            fantasy_counter += 1
+        elif movie["genre"] == "Action":
+            action_counter += 1
+        elif movie["genre"] == "Intrigue":
+            intrigue_counter += 1
+    if horror_counter > fantasy_counter and horror_counter > action_counter and horror_counter > intrigue_counter:
+        return "Horror"
+    if fantasy_counter > horror_counter and fantasy_counter > action_counter and fantasy_counter > intrigue_counter:
+        return "Fantasy"
+    if action_counter > horror_counter and action_counter > fantasy_counter and action_counter > intrigue_counter:
+        return "Action"
+    if intrigue_counter > horror_counter and intrigue_counter > fantasy_counter and intrigue_counter > action_counter:
+        return "Intrigue"
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+# finds the movies the user has watched AND their friends have not
+def get_unique_watched(user_data):
+    friend_movies = set()
+    user_movies = set()
+    unique_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.add(movie["title"])
+    for movie in user_data["watched"]:
+        user_movies.add(movie["title"])
+    unique_titles = user_movies - friend_movies
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_titles:
+            unique_watched.append(movie)
+    return unique_watched
+
+# finds the movies the user's friends have watched that they have not
+def get_friends_unique_watched(user_data):
+    friend_movies = set()
+    user_movies = set()
+    unique_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.add(movie["title"])
+    for movie in user_data["watched"]:
+        user_movies.add(movie["title"])
+    unique_titles = friend_movies - user_movies
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in unique_titles:
+                if movie not in unique_watched:
+                    unique_watched.append(movie)
+    return unique_watched

```

Comment:
There's some shared logic I'm seeing between the two functions in Wave 3, which is usually a sign that some code could be pulled into a helper function.